pygen/pygen_src/riscv_asm_program_gen.py

```diff
@@ -704,9 +745,66 @@ def gen_test_file(self, test_name):
         file.close()
         logging.info("%0s is generated", test_name)
 
-    def gen_signature_handshake(self, instr, signature_type, core_status = "INITIALIZED",
-                                test_result = "TEST_FAIL", csr = "MSCRATCH", addr_label = ""):
-        pass
+    def gen_signature_handshake(self, instr, signature_type,
+                                core_status = core_status_t.INITIALIZED,
+                                test_result = test_result_t.TEST_FAIL,
+                                csr = privileged_reg_t.MSCRATCH,
+                                addr_label = ""):
+        if cfg.require_signature_addr:
+            string = []
+            string = ("li x{}, {}".format(cfg.gpr[1].value, hex(cfg.signature_addr)))
+            instr.extends(string)
+            # A single data word is written to the signature address.
+            # Bits [7:0] contain the signature_type of CORE_STATUS, and the upper
+            # XLEN-8 bits contain the core_status_t data.
+            if signature_type == signature_type_t.CORE_STATUS:
+                string.extend(("li x{}, {}".format(cfg.gpr[0].value, hex(core_status.value)),
+                               "slli x{}, x{}, 8".format(cfg.gpr[0].value, cfg.gpr[0].value),
+                               "addi x{}, x{}, {}".format(cfg.gpr[0].value, cfg.gpr[0].value,
+                                                          hex(signature_type.value)),
+                               "sw x{}, 0(x{})".format(cfg.gpr[0].value, cfg.gpr[1].value)))
+                instr.extend(string)
+            # A single data word is written to the signature address.
+            # Bits [7:0] contain the signature_type of TEST_RESULT, and the upper
+            # XLEN-8 bits contain the test_result_t data.
+            elif signature_type == test_result_t.TEST_RESULT:
+                string.extend(("li x{}, {}".format(cfg.gpr[0].value, hex(test_result.value)),
+                               "slli x{}, x{}, 8".format(cfg.gpr[0].value, cfg.gpr[0].value),
+                               "addi x{}, x{}, {}".format(cfg.gpr[0], cfg.gpr[0],
+                                                          hex(signature_type.value)),
+                               "sw x{}, 0(x{})".format(cfg.gpr[0].value, cfg.gpr[1].value)))
+                instr.extend(string)
```

Comment:
I believe there is a tiny bug here: logically, we extend the instr once in line 756, then in the next "if else" statements, we extend the string, and again extend the instr. This way, I believe we are appending line 755 multiple times which could get problematic. Does that make sense?

Reply:
yes, It makes sense now.
I have removed the string list from here and directly did the extend with instr only.